text/0079-cdk-2.0.md

```diff
@@ -0,0 +1,632 @@
+major version, immediately stop supporting the previous version, and leave the onus on the user to "catch up".
+
+Therefore, a major part of the motivation to perform this work is to establish a framework for releasing new major
+versions of the cdk in a controlled and standardized way. Establishing standards and tools to automatically handle
+deprecations, feature flag resets, and pre-release versioning will allow the core team and contributors to leverage
+these to better control breakage. The path to introduce a breaking change is then a lot more clear.
+
+First we ask, "can this be implemented behind a feature flag"? If so, that's usually the answer. Build tooling tells us
+when feature flags can be removed and warns users when it has happened. Migration path to new behavior has to be
+considered and should be documented clearly at the time the flag is created. This goes in the migration guide of the
+next major version.
+
+If it can't be implemented behind a feature flag, we ask "can a new code path be defined and the old one deprecated?".
+Once again, migration path is considered and documented and it goes into the next major version migration guide.
+
+If the migration path is too painful in either situation, can it be reduced somehow? Can it be automated? etc...
+
+The desire to move forward with the change in packaging strategy has brought the need for these tools and procedures to
+the forefront.
+
+## Alternatives
+
+The most notable alternative to releasing a new major version, is basically to not. This could be done a couple of ways.
+
+The first is not moving forward with the packaging change at all. For the argument against that, see the
+[MonoCDK RFC](https://github.com/aws/aws-cdk-rfcs/issues/6).
+
+The second is to simply release the new module, `aws-cdk-lib` locked to the same `1.x.x` version that the other modules
+are currently locked to. In fact, that is the first stage of the release process, its just under the name
+`monocdk-experiment`. This was a low cost way for the dev team, contributors, and users to experiment with the new
+structure before investing too much towards it.
+
+Regardless, the point of using this change to push towards a v2 release, is that the single package structure is "the
+new cdk". Meaning, the long term plan is to only release the `aws-cdk-lib` package and no longer publish the other
+`@aws-cdk/` namespaced ones. The [MonoCDK RFC](https://github.com/aws/aws-cdk-rfcs/issues/6) has more info on the
+problems currently experienced by users caused by the current packaging strategy and these problems will always exist
+unless we stop supporting the `@aws-cdk/` packages.
+
+Since we want to eventually supplant the current module structure with the new one, semver says we have to do it with
+the release of a major version.
+
+# Adoption Strategy
+
+All users will perform the following to adopt CDK v2
+
```

Comment:
What's the feasibility of a migration tool to help customers update their dependencies and imports?

Reply:
All of this is *easy* if you consider **one** language (**TypeScript**/**Javascript**). Things become **much** hairier once you start to think about **Python**, **C#**, **Java**, ... I think if we go into tooling we should put ourselves in the position to offer an equally managed experience for **all** languages (and I don't think this is realistic).